data/collector.go

```diff
@@ -36,36 +37,41 @@ type Collector interface {
 }
 
 type collector struct {
-	clock             clock.Clock
-	queue             chan *v1.SensorData
-	interval          time.Duration
-	params            map[string]*anypb.Any
-	lock              *sync.Mutex
-	logger            golog.Logger
-	backgroundWorkers sync.WaitGroup
-	cancelCtx         context.Context
-	cancel            context.CancelFunc
-	captureFunc       CaptureFunc
-	closed            bool
-	target            datacapture.BufferedWriter
+	clock          clock.Clock
+	captureResults chan *v1.SensorData
+	captureErrors  chan error
+	interval       time.Duration
+	params         map[string]*anypb.Any
+	lock           *sync.Mutex
+	logger         golog.Logger
+	captureWorkers sync.WaitGroup
+	logRoutine     sync.WaitGroup
+	cancelCtx      context.Context
+	cancel         context.CancelFunc
+	captureFunc    CaptureFunc
+	closed         *atomic.Bool
```

Comment:
Can you explain the motivation for changing this to be atomic? Can `Close()` can be called on the collector from a number of different goroutines?

Reply:
Just looked through its usages, and nope looks like Close cannot be called from different routines anywhere (everywhere it's called in builtin, the builtin lock is held). So went ahead and changed this to a normal bool